File: e189/python/solution.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def firstMissingPositive(self, A):
        # write your code here
        size = len(A)
        B = [0] * size

        for i in range(size):
            if A[i] >= 1 and A[i] <= size:
                B[A[i]-1] = A[i]

        for i in range(size):
            if B[i] == 0:
                return i+1

        return size+1


def main():
    A = [1,2,0]
    print Solution().firstMissingPositive(A)

if __name__ == '__main__':
    main()

File: e189/java/Solution.java
public class Solution {
    /**    
     * @param A: an array of integers
     * @return: an integer
     */
    public int firstMissingPositive(int[] A) {
        // write your code here   
        int len = A.length;
        int[] flag = new int[len];
        for (int i=0; i < len; i++) {
            if (A[i] >= 1 & A[i] <= len) {
                flag[A[i] - 1] = 1;
            }
        }
        
        for (int j=0; j < len; j++) {
            if (flag[j] == 0) {
                return j+1;
            }
        }
        return len+1;
    }
}
